gordon/router.py

```diff
@@ -90,20 +95,45 @@ def _get_next_phase(self, event_msg):
             next_phase = 'cleanup'
         return next_phase
 
-    async def _route(self, event_msg, next_phase=None):
-        if not interfaces.IEventMessage.providedBy(event_msg):
-            msg = (f'Ignoring message "{event_msg.msg_id}". Does not correctly'
-                   ' implement `IEventMessage`.')
-            logging.warn(msg)
-            return
+    async def _update_messages_in_flight(self):
+        in_flight_count = len(self._messages_in_flight)
+        await self.metrics.set(
+            'router-messages-in-flight', value=in_flight_count)
 
-        if not next_phase:
+    async def _add_message_in_flight(self, event_msg):
+        if event_msg.msg_id not in self._messages_in_flight:
+            context = {'unit': 'seconds'}
+            timer = self.metrics.timer(
+                'router-message-flight-duration', context=context)
+            await timer.start()
+            self._messages_in_flight[event_msg.msg_id] = timer
+
+        await self._update_messages_in_flight()
+
+    async def _remove_message_in_flight(self, event_msg):
+        if event_msg.phase in self.FINAL_PHASES:
+            timer = self._messages_in_flight.pop(event_msg.msg_id)
+            await timer.stop()
+
+        await self._update_messages_in_flight()
+
+    async def _route(self, event_msg, force_cleanup=False):
+        if force_cleanup:
+            next_phase = 'cleanup'
+        else:
```

Comment:
I know this boolean check could be handled a bit cleaner, but I wanted to be super explicit.